seq2simulate/run_make_proviral_data.py

```diff
@@ -0,0 +1,447 @@
+        or sam_line['read_start'] >= seq_diffs[0] and sam_line['read_end'] > seq_diffs[0]:
+        return True
+    return False
+
+def _is_stopcodon(sam_line, seq_diffs):
+    if sam_line['read_start'] <= seq_diffs[0] and sam_line['read_end'] >= seq_diffs[0]:
+        return True
+    return False
+
+def _is_inversion(sam_line, seq_diffs):
+    if seq_diffs[0] <= sam_line['read_start'] <= seq_diffs[1] \
+        or seq_diffs[0] <= sam_line['read_end'] <= seq_diffs[1] \
+        or sam_line['read_start'] <= seq_diffs[0] and sam_line['read_end'] >= seq_diffs[1] \
+        or sam_line['read_start'] >= seq_diffs[0] and sam_line['read_end'] <= seq_diffs[1]:
+        return True
+    return False
+
+
+def sample_fastq(n_reads, fastq1=None, fastq2=None):
+
+    """
+    Returns a random sample of a list form of a FASTQ file (output of
+    custom_generators.parse_fastq())
+
+    Args:
+        n_reads: number of reads to sample
+        fastq1/2: open FASTQ file objects (paired-end data has forward and
+        reverse files)
+    """
+    paired_end = False
+    if fastq1 is None and fastq2 is None:
+        raise TypeError('At least one of fastq1 and fastq2 must not be None.')
+    elif fastq1 is None and fastq2 is not None:
+        fastq1, fastq2 = fastq2, fastq1
+    elif fastq1 is not None and fastq2 is not None:
+        paired_end = True
+
+    if paired_end:
+        if len(fastq1) != len(fastq2):
+            raise ValueError('FASTQ forward and reverse files must have the same lengths.')
+
+    fastq_sample1 = []
+    if paired_end:
+        fastq_sample2 = []
+
+    while len(fastq_sample1) < n_reads:
+        
+        i = random.randint(0, len(fastq1) - 1)
+        proviral_read1 = list(fastq1.pop(i))
+        fastq_sample1.append(proviral_read1)
+        if paired_end:
+            proviral_read2 = list(fastq2.pop(i))
+            fastq_sample2.append(proviral_read2)
+
+    if paired_end:    
+        return fastq_sample1, fastq_sample2
+    else:
+        return fastq_sample1
+
+def _parse_sam_line(read_id, sam_file, paired_end=False):
+    """
+    Parse a single SAM file line (or pair of lines for paired-end data).
+
+    Args:
+        read_id: id string of FASTQ read
+        sam_file: output of custom_generators.parse_sam() 
+        paired_end: is this SAM file for paired-end data
+    """
+    if paired_end:
+        seq_id_1 = sam_file[read_id][0][1] 
+        seq_id_2 = sam_file[read_id][1][1] 
+        if seq_id_1 != seq_id_2:
+            raise ValueError('read_id does not reference paired reads from same sequence: {} {}'.format(seq_id_1, seq_id_2))
+        if int(sam_file[read_id][0][7]) < 0 and int(sam_file[read_id][1][7]) > 0:
+            sam_file[read_id] = sam_file[read_id][::-1]
+        elif int(sam_file[read_id][0][7]) > 0 and int(sam_file[read_id][1][7]) > 0 or int(sam_file[read_id][0][7]) < 0 and int(sam_file[read_id][1][7]) < 0:
```

Comment:
or isn't this possibly just an "else", because you've already covered all the other cases?  This number is never zero due to the joys of 1-numbering in sam files if I understand.